### ISSUE TITLE
WARN gpgui::portal_connector] Failed to connect to the portal: Prelogin error: Invalid prelogin response

### ISSUE THREAD

**Issue by keisner** [yuezk/GlobalProtect-openconnect#357]
I continue to get the error Invalid prelogin response. I have tried changing most of the settings in the client, along with the "Versioning". Please see logs below:

Location: /home/cooljoe51/.local/share/gpclient/gpclient.log
[2024-05-06T15:46:47Z WARN gpgui::portal_connector] Failed to connect to the portal: Prelogin error: Invalid prelogin response
[2024-05-06T16:06:55Z INFO gpgui::handlers::subscription] Sending the init event to client: settings
[2024-05-06T16:06:55Z INFO gpgui::handlers::subscription] Sent the init event to client: settings
[2024-05-06T16:12:14Z INFO gpgui::portal_connector] Connecting to the portal: v**********m...
[2024-05-06T16:12:14Z INFO gpgui::portal_connector] Connecting the portal as a gateway...
[2024-05-06T16:12:14Z INFO gpgui::portal_connector] Gateway prelogin, gateway: v**********m...
[2024-05-06T16:12:14Z INFO gpapi::portal::prelogin] Gateway prelogin with user_agent: PAN GlobalProtect/5.2.6 (Microsoft Windows 11 Pro , 64-bit)
[2024-05-06T16:12:14Z WARN gpapi::portal::prelogin] Parse response error, response: <?xml version="1.0" encoding="UTF-8" ?>
[2024-05-06T16:12:14Z WARN gpgui::portal_connector] Failed to connect to the portal: Prelogin error: Invalid prelogin response
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Connecting to the portal: v**********m...
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Trying to connect the gateway directly...
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Failed to connect the gateway directly: No portal connection found
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Trying to connect portal with cached credential...
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Failed to connect portal with cached credential: No cached credential found for the portal
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Trying to connect the portal with prelogin...
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Performing portal prelogin...
[2024-05-06T17:12:22Z INFO gpapi::portal::prelogin] Portal prelogin with user_agent: PAN GlobalProtect/5.2.6 (Microsoft Windows 11 Pro , 64-bit)
[2024-05-06T17:12:22Z WARN gpapi::portal::prelogin] Parse response error, response: <?xml version="1.0" encoding="UTF-8" ?>
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Failed to connect the portal with prelogin: Prelogin error: Invalid prelogin response
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Trying to connect the portal as a gateway...
[2024-05-06T17:12:22Z INFO gpgui::portal_connector] Gateway prelogin, gateway: v**********m...
[2024-05-06T17:12:22Z INFO gpapi::portal::prelogin] Gateway prelogin with user_agent: PAN GlobalProtect/5.2.6 (Microsoft Windows 11 Pro , 64-bit)
[2024-05-06T17:12:22Z WARN gpapi::portal::prelogin] Parse response error, response: <?xml version="1.0" encoding="UTF-8" ?>
[2024-05-06T17:12:22Z WARN gpgui::portal_connector] Failed to connect to the portal: Prelogin error: Invalid prelogin response
[2024-05-06T18:34:59Z INFO gpgui::handlers::subscription] Sending the init event to client: settings
[2024-05-06T18:34:59Z INFO gpgui::handlers::subscription] Sent the init event to client: settings

**Comment by yuezk** [comment]
It seems that the portal pre-login endpoint returns an invalid XML response. What's the response of `https://<vpn_portal>/global-protect/prelogin.esp`.

**Comment by keisner** [comment]
Looks like Success:

<prelogin-response>
<status>Success</status>
<ccusername/>
<autosubmit>false</autosubmit>
<msg/>
<newmsg/>
<authentication-message>Enter login credentials</authentication-message>
<username-label/>
<password-label/>
<panos-version>1</panos-version>
<saml-default-browser>yes</saml-default-browser>
<auth-api>no</auth-api>
<region>US</region>
</prelogin-response>

**Comment by yuezk** [comment]
Hi @keisner can you help try this to see if it works for you.

1. Click the top left menu, select `Clear Credentials`
2. Click the icon of the portal input
   ![image](https://github.com/yuezk/GlobalProtect-openconnect/assets/3297602/8262fd15-7efa-4f17-ae3f-782f25fe30a7)
3. Select the `Portal Server`
   ![image](https://github.com/yuezk/GlobalProtect-openconnect/assets/3297602/4ea9ddc9-9537-4c72-878c-834997936fb1)
5. Connect it again.

Please provide the full logs if it still fails.

**Comment by keisner** [comment]
I have done this a few times and switched portals. This did not work at allOn May 7, 2024, at 11:56 AM, Kevin Yue ***@***.***> wrote:﻿
Hi @keisner can you help try this to see if it works for you.

Click the top left menu, select Clear Credentials
Click the icon of the portal input
image.png (view on web)
Select the Portal Server
image.png (view on web)
Connect it again.

Please provide the full logs if it still fails.

**Comment by yuezk** [comment]
Have this client ever worked for you?

**Comment by keisner** [comment]
Yes. When I was version 9On May 7, 2024, at 12:12 PM, Kevin Yue ***@***.***> wrote:﻿
Have this client ever worked for you?

**Comment by yuezk** [comment]
> Yes. When I was version 9

What do you mean by version 9? Version 2.0.0-beta9?

**Comment by keisner** [comment]
Panos9

**Comment by yuezk** [comment]
> Looks like Success:
> 
> ```
> <prelogin-response>
> <status>Success</status>
> <ccusername/>
> <autosubmit>false</autosubmit>
> <msg/>
> <newmsg/>
> <authentication-message>Enter login credentials</authentication-message>
> <username-label/>
> <password-label/>
> <panos-version>1</panos-version>
> <saml-default-browser>yes</saml-default-browser>
> <auth-api>no</auth-api>
> <region>US</region>
> </prelogin-response>
> ```

Hi @keisner I can reproduce a similar error using this prelogin response.

Have you tried using the official client on macOS/Windows to connect this portal? Do they work for you?

**Comment by keisner** [comment]
The windows one does work. I even tested a known working version in the
configuration.

**Comment by yuezk** [comment]
I guess the official client asked you to input the username and password. Is that right? If yes, the reason might be that the prelogin response doesn't have values for `username-label` and `password-label`, this could fail parsing the prelogin response. I can use the fallback value for them.

**Comment by keisner** [comment]
Yes you are correct. My username and password had to be entered.

Is there anything I can attempt to do on my end to assist in testing?

**Comment by yuezk** [comment]
I will provide a test package later today, you can give it a try.

**Comment by keisner** [comment]
Okay thanks!

**Comment by yuezk** [comment]
@keisner, you can try the test package from https://github.com/yuezk/GlobalProtect-openconnect/releases/tag/snapshot. Remember to uninstall the `globalprotect-openconnect` package before installing the test package.

**Comment by keisner** [comment]
I tried with no success:

[2024-05-09T14:18:36Z INFO gpgui::handlers::subscription] Sent the init
event to client: main
[2024-05-09T14:18:41Z INFO gpgui::handlers::subscription] Sending the init
event to client: settings
[2024-05-09T14:18:41Z INFO gpgui::handlers::subscription] Sent the init
event to client: settings
[2024-05-09T14:52:40Z INFO gpgui::handlers::subscription] Sending the init
event to client: main
[2024-05-09T14:52:40Z INFO gpgui::handlers::subscription] Sent the init
event to client: main
[2024-05-09T14:53:01Z INFO gpgui::handlers::subscription] Sending the init
event to client: settings
[2024-05-09T14:53:01Z INFO gpgui::handlers::subscription] Sent the init
event to client: settings
[2024-05-09T15:06:42Z INFO gpgui::handlers::subscription] Sending the init
event to client: main
[2024-05-09T15:06:42Z INFO gpgui::handlers::subscription] Sent the init
event to client: main
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Connecting to the
portal: v**********m...
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Trying to connect the
gateway directly...
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Failed to connect the
gateway directly: No portal connection found
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Trying to connect
portal with cached credential...
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Failed to connect
portal with cached credential: No cached credential found for the portal
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Trying to connect the
portal with prelogin...
[2024-05-09T15:07:13Z INFO gpgui::portal_connector] Performing portal
prelogin...
[2024-05-09T15:07:13Z INFO gpapi::portal::prelogin] Portal prelogin with
user_agent: PAN GlobalProtect/5.2.6 (Microsoft Windows 11 Pro , 64-bit)
[2024-05-09T15:07:20Z WARN gpapi::portal::prelogin] Parse response error,
response: <?xml version="1.0" encoding="UTF-8" ?>
[2024-05-09T15:07:20Z INFO gpgui::portal_connector] Failed to connect the
portal with prelogin: Prelogin error: Invalid prelogin response
[2024-05-09T15:07:20Z INFO gpgui::portal_connector] Trying to connect the
portal as a gateway...
[2024-05-09T15:07:20Z INFO gpgui::portal_connector] Gateway prelogin,
gateway: v**********m...
[2024-05-09T15:07:20Z INFO gpapi::portal::prelogin] Gateway prelogin with
user_agent: PAN GlobalProtect/5.2.6 (Microsoft Windows 11 Pro , 64-bit)
[2024-05-09T15:07:20Z WARN gpapi::portal::prelogin] Parse response error,
response: <?xml version="1.0" encoding="UTF-8" ?>
[2024-05-09T15:07:20Z WARN gpgui::portal_connector] Failed to connect to
the portal: Prelogin error: Invalid prelogin response
[2024-05-09T15:07:42Z INFO gpgui::handlers::subscription] Sending the init
event to client: settings
[2024-05-09T15:07:42Z INFO gpgui::handlers::subscription] Sent the init
event to client: settings

**Comment by yuezk** [comment]
Hi @keisner would you mind send me your portal address via email k3vinyue@gmail.com? So I can investigate it. The credentials are not needed.

**Comment by keisner** [comment]
I emailed you

**Comment by yuezk** [comment]
Hi @keisner, it works for me with your VPN portal. I'm afraid this is related to the installation process. What's your OS and how did you install the old client?

Anyway, you can try these steps to see if it works for you:
1. Quit the client
2. Remove the package from your system.
3. Run `sudo rm -f /usr/bin/gpgui` to ensure the old `gpgui` binary is removed.
4. Install the test package from https://github.com/yuezk/GlobalProtect-openconnect/releases/tag/snapshot
5. Launch the GUI client, which will download the snapshot `gpgui` dynamically.
6. Check if it works for you.

**Comment by yuezk** [comment]
Glad it worked for you. I will release the fix in v2.2.2

**Comment by keisner** [comment]
Thanks so much!

**Comment by yuezk** [comment]
Released in 2.3.0, closing.
